go-controller/pkg/util/pod_annotation.go

```diff
@@ -300,14 +301,37 @@ func GetPodCIDRsWithFullMask(pod *v1.Pod, nInfo NetInfo) ([]*net.IPNet, error) {
 	return ips, nil
 }
 
-// GetPodIPsOfNetwork returns the pod's IP addresses, first from the OVN annotation
-// and then falling back to the Pod Status IPs. This function is intended to
-// also return IPs for HostNetwork and other non-OVN-IPAM-ed pods.
+// GetPodIPsOfNetwork returns the pod's IP addresses.
+// If the network is default type, then retrieve it from Pod Status IPs.
+// If the network is secondary and also used as primary network for the pod, then retrieve
+// IPs from "k8s.ovn.org/pod-networks" pod annotation.
+// If the network is only secondary, then retrieve IPs from the k8s.ovn.org/pod-networks
+// annotation based on NetworkSelectionElement.
+// This function is intended to also return IPs for HostNetwork and other non-OVN-IPAM-ed pods.
 func GetPodIPsOfNetwork(pod *v1.Pod, nInfo NetInfo) ([]net.IP, error) {
-	if nInfo.IsSecondary() {
-		return SecondaryNetworkPodIPs(pod, nInfo)
+	if !nInfo.IsSecondary() || PodWantsHostNetwork(pod) {
+		return DefaultNetworkPodIPs(pod)
+	}
+	if nInfo.IsPrimaryNetwork() {
+		nadList := nInfo.GetNADs()
+		for _, nad := range nadList {
+			userNet, err := UnmarshalPodAnnotation(pod.Annotations, nad)
+			if err != nil && strings.Contains(err.Error(), "no ovn pod annotation for network") {
+				// NAD belongs to same network, but serving different namespace. so skip processing it.
```

Comment:
Some things here:

- Un-marshaling the annotation for each NAD is inefficient
- Why would be a pod be annotated with a NAD of a different namespace?
- Use a well defined error. We should we using `newAnnotationNotSetError`, we are probably not doing it for historic reasons but I don't see why.

Reply:
> * Un-marshaling the annotation for each NAD is inefficient

do we have any other option other than parsing netwoks annotation ?

> * Why would be a pod be annotated with a NAD of a different namespace?

This function may be invoked with netInfo object which is global object for multiple namespaces, hence needed logic to fetch right NAD name for the namespace.

> * Use a well defined error. We should we using `newAnnotationNotSetError`, we are probably not doing it for historic reasons but I don't see why.

yes, using `newAnnotationNotSetError` for this purpose now, didn't see any issue during the tests.

Reply:
> > * Un-marshaling the annotation for each NAD is inefficient
> 
> do we have any other option other than parsing netwoks annotation ?

You can parse once getting a map of NADs to IPs. Instead of parsing the full annotation for each NAD.

> > * Why would be a pod be annotated with a NAD of a different namespace?
> 
> This function may be invoked with netInfo object which is global object for multiple namespaces, hence needed logic to fetch right NAD name for the namespace.

So why not use `GetActiveNetworkForNamespace` which will give you the NADs for the namespace? This is related to the previous comment, you are parsing the annotation for each NAD when many of them might not even be relevant.

> 
> > * Use a well defined error. We should we using `newAnnotationNotSetError`, we are probably not doing it for historic reasons but I don't see why.
> 
> yes, using `newAnnotationNotSetError` for this purpose now, didn't see any issue during the tests.

Did you check that this is ok for previously existing uses of `UnmarshalPodAnnotation`? While it should be fine, we need to be careful.

Reply:
> So why not use `GetActiveNetworkForNamespace` which will give you the NADs for the namespace? This is related to the previous comment, you are parsing the annotation for each NAD when many of them might not even be relevant.

Meaning, GetActiveNetworkForNamespace should give you a netinfo with a single NAD because there can only be one primary NAD per namespace, so the issue wouldn't be as important.